Mailbot - new violation handler

micmarty opened the pull request:
Read all sections carefully :tea: 

## Features
* Mailbot sends email every `interval` minutes (each email address is timed separately)
* Admin can also get emails about violations (useful, when intruder does not have TH account or uses invalid/fake email address)
* Admin and intruder can get different email bodies
* There is a body templating mechanism that supports HTML and is fully customizable via `mailbot_config.ini`

## Preview
### Email to intruder
![image](https://user-images.githubusercontent.com/12485656/51789201-af2ff200-2186-11e9-946f-16d993835868.png)

### Email to admin:
![image](https://user-images.githubusercontent.com/12485656/51789016-b3f3a680-2184-11e9-9708-b99d3102931f.png)

### Email HTML templating:
![image](https://user-images.githubusercontent.com/12485656/51789022-bce47800-2184-11e9-9abf-d85e7cc15b6e.png)


## Configuration
* Uses separate config file - `mail_config.ini`
![image](https://user-images.githubusercontent.com/12485656/51788898-28c5e100-2183-11e9-9448-ced590f59366.png)
* `mail_config.ini` will be always recreated (identically as other config files)
* Mailbot's SMTP credentials must be defined by 2 environment variables (but the names can be customized in config file)
* User will know if these env vars are undeclared
![image](https://user-images.githubusercontent.com/12485656/51788895-1ba8f200-2183-11e9-9b65-fb7df5e14dd4.png)

## Manual testing scenarios
- Make clean install of `tensorhive` (remove everything under `~/.config/Tensorhive/`)
- Remember to fill in `hosts_config.ini`
- Notice warning about undeclared env variables,
```bash
export TENSORHIVE_MAILBOT_EMAIL="ask @micmarty"
export TENSORHIVE_MAILBOT_PASSWORD="ask @micmarty"
```
- Create accounts:
`tensorhive create user`
- create admin account
- create intruder account
- Make reservation as admin
- Run GPU process as intruder
- Check intruder's inbox (and optionally admin's - `notify_admin = yes`)

### More:
- experiment with `mailbot_config.ini` to find bugs
- look for edge cases

## Notes
@roscisz 
Code used to be clean at some point, now it looks clunky, especially for me/you.
I'll do a hard refactor, because that code will be hard to maintain/extend later.
I have a concept in my mind already, so it's a matter of time and priorities.
What's your opinion?

roscisz commented on tensorhive/models/User.py:
```diff
@@ -25,6 +25,7 @@ class User(CRUDModel, Base):
     __tablename__ = 'users'
     id = Column(Integer, primary_key=True, autoincrement=True)
     username = Column(String(40), unique=True, nullable=False)
+    email = Column(String(40), unique=True, nullable=False)
```
Does the e-mail have to be unique? Maybe I would like to have two accounts (for example one admin and one normal for testing).

micmarty replied:
Ok, u're right

roscisz commented on tensorhive/models/User.py:
```diff
@@ -74,6 +75,11 @@ def validate_username(self, key, username):
 
+    @validates('email')
+    def validate_username(self, key, email):
+        log.warning('TODO Validate email correctness!')
```
TODO

roscisz commented on tensorhive/core/services/ProtectionService.py:
```diff
@@ -174,8 +185,20 @@ def do_run(self):
+                    'UUID': uuid,
+                    'HOSTNAME': hostname
+                }
+                self.violation_handlers[1].trigger_action(violation_data)
```
If i change notify_on_pty in config.ini, there is no violation_handlers[1].

BTW shouldn't the trigger_action method have the same interface for all handlers? If the parameters differ, maybe we could use a map of params?

micmarty replied:
I was thinking exactly the same (about common interface), but it turned out that my initial implementation was just bad (it requires ssh `connections` argument which is useless for mailer - short-sighted vision).

And because I didn't want to screw up current code.... that's the workaround which I'm feeling kinda ashamed for.

I'll try to make it workin'. :cactus:

micmarty replied:
It took me a while, but yeah... there it is (refactor was necessary indeed, see changes in `MessageSendingBehaviour`):
![image](https://user-images.githubusercontent.com/12485656/51948336-64350980-2428-11e9-8f37-f53b1e12348b.png)

roscisz commented on tensorhive/config.py:
```diff
@@ -178,7 +197,39 @@ class PROTECTION_SERVICE:
     UPDATE_INTERVAL = config.getfloat(section, 'update_interval', fallback=2.0)
     NOTIFY_ON_PTY = config.getboolean(section, 'notify_on_pty', fallback=True)
-
+    NOTIFY_VIA_EMAIL = config.getboolean(section, 'notify_on_pty', fallback=True)
```
notify_on_pty should be changed to notify_via_email

micmarty replied:
Should this option be disabled by default?

roscisz commented on tensorhive/config.py:
```diff
@@ -178,7 +197,39 @@ class PROTECTION_SERVICE:
+
+    # FIXME Not sure if this should be required
+    section = 'smtp'
+    SMTP_LOGIN_ENV = mailbot_config.get(section, 'email_env_var')
```
Why do we specify email_env_var in config instead of simply specifying email and password in config?

micmarty replied:
Email address: maybe yes..., but I have mixed feelings about the password...
I thought that in order to be concise these should be treated equally.

My first experience with such dilemma was about 2 years ago with Docker and the answer was:
![image](https://user-images.githubusercontent.com/12485656/51943123-03520500-2419-11e9-9c62-a7264133c645.png)

All in all, we can keep the password in `.ini` files as long as we're sure that noone, besides admin, can open this file (admin should take care of it by himself).

roscisz replied:
Maybe we could keep the password in the mailbot config, but make sure to protect the file while creating it:
chmod go-r ~/.config/TensorHive/mailbot_config.ini

roscisz commented on tensorhive/core/utils/mailer.py:
```diff
@@ -0,0 +1,91 @@
+        try:
+            body = self.template.format(
+                hostname=data['HOSTNAME'],
+                gpu_name='TODO',
```
TODO?

micmarty replied:
Implemented:
![image](https://user-images.githubusercontent.com/12485656/51943715-86c02600-241a-11e9-937a-b92e3a4da294.png)

micmarty commented on tensorhive/send_mail.py:
```diff
@@ -0,0 +1,24 @@
+body = "Alert!"
+msg.attach(MIMEText(body, 'plain'))
+ 
+server = smtplib.SMTP('mail.linux.pl', 587)
```
This was my toy file. It will be removed :(

micmarty commented on npm-debug.log:
```diff
@@ -0,0 +1,66 @@
+0 info it worked if it ends with ok
```
> Adding the non-null e-mail field in User model makes it impossible to add user through Webapp:
> 
> `NOT NULL constraint failed: users.email with <User id=None, username`
> 
> We should either remove the not-null constraint or add appropriate changes to webapp in this PR. I suggest latter :)
Done 
![image](https://user-images.githubusercontent.com/12485656/51942612-d05b4180-2417-11e9-8924-703b099aed77.png)

![image](https://user-images.githubusercontent.com/12485656/51942586-bfaacb80-2417-11e9-9428-3f632adfc621.png)

micmarty commented on tensorhive/app/web/dev/package-lock.json:
```diff
@@ -0,0 +1,13743 @@
+{
```
Yes, but it should be there. Source:
![image](https://user-images.githubusercontent.com/12485656/51942863-655e3a80-2418-11e9-9556-946299c33385.png)

micmarty replied:
What do we do then?

micmarty replied:
Decision at team meeting: file should stay

micmarty commented:
All "Resolved conversations" are fixed in this commit: 
https://github.com/roscisz/TensorHive/pull/142/commits/d3c91ca2e81fb8ddb68690ef85e797d02fbb045b

roscisz commented on tensorhive/mailbot_config.ini:
```diff
@@ -0,0 +1,58 @@
+[smtp]
+email_env_var = TENSORHIVE_MAILBOT_EMAIL
+password_env_var = TENSORHIVE_MAILBOT_PASSWORD
+smtp_server = smtp.gmail.com
```
There should be only a description "mail server hostname", instead of a specific address

micmarty replied:
![image](https://user-images.githubusercontent.com/12485656/52073710-78901800-2588-11e9-86c3-3866a6c0066c.png)
`mailbot_config.ini`
![image](https://user-images.githubusercontent.com/12485656/52073717-7ded6280-2588-11e9-9a4e-99c8edff7e75.png)
`MAILBOT` config class:
![image](https://user-images.githubusercontent.com/12485656/52073796-b725d280-2588-11e9-9174-57519cd6ace4.png)

micmarty replied:
@roscisz On our last meeting (31.01.2019) you suggested that admin should get emails less frequently than intruder (e.g. just once every `X` intruder's notifications, or once per unprivileged process, etc.)

I think it'll be better to **introduce this feature later** (with bootstrap or something).
I created more detailed issue for this problem: https://github.com/roscisz/TensorHive/issues/145

roscisz commented:
I suppose we should keep track of the pid that is violating the reservation, so that we send only 1 warning email per pid. We don't want the admin to be spammed after a night of violations.

micmarty commented:
* Fixed user update in API controller (was not adjusted to handle email attribute from request)
* Resolved conflict with develop
* Manually tested
* Added new issue (found out by @roscisz) #149 

For convenience, comment out `log.debug()` statements in these files
![image](https://user-images.githubusercontent.com/12485656/53429425-f19a5800-39ec-11e9-8474-1e08856fda8f.png)

It should be ready to merge :)
